pkg/spanconfig/spanconfigsqltranslator/protectedts_state_reader_test.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package spanconfigsqltranslator_test
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"testing"
+
+	_ "github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl"
+	_ "github.com/cockroachdb/cockroach/pkg/ccl/partitionccl"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobsprotectedts"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/protectedts/ptpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigsqltranslator"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+func TestProtectedTimestampStateReader(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	mkRecordAndAddToState := func(state *ptpb.State, ts hlc.Timestamp, target *ptpb.Target) {
+		recordID := uuid.MakeV4()
+		rec := jobsprotectedts.MakeRecord(recordID, int64(1), ts, nil, /* deprecatedSpans */
+			jobsprotectedts.Jobs, target)
+		state.Records = append(state.Records, *rec)
+	}
+
+	protectSchemaObject := func(state *ptpb.State, ts hlc.Timestamp, ids []descpb.ID) {
+		mkRecordAndAddToState(state, ts, ptpb.MakeSchemaObjectsTarget(ids))
+	}
+
+	protectCluster := func(state *ptpb.State, ts hlc.Timestamp) {
+		mkRecordAndAddToState(state, ts, ptpb.MakeClusterTarget())
+	}
+
+	protectTenants := func(state *ptpb.State, ts hlc.Timestamp, ids []roachpb.TenantID) {
+		mkRecordAndAddToState(state, ts, ptpb.MakeTenantsTarget(ids))
+	}
+
+	// Create some ptpb.State and then run the ProtectedTimestampStateReader on it
+	// to ensure it outputs the expected records.
+	state := &ptpb.State{}
+	protectSchemaObject(state, hlc.Timestamp{WallTime: 1}, []descpb.ID{56})
+	protectSchemaObject(state, hlc.Timestamp{WallTime: 2}, []descpb.ID{56, 57})
+	protectCluster(state, hlc.Timestamp{WallTime: 3})
+	protectTenants(state, hlc.Timestamp{WallTime: 4}, []roachpb.TenantID{roachpb.MakeTenantID(1)})
+	protectTenants(state, hlc.Timestamp{WallTime: 5}, []roachpb.TenantID{roachpb.MakeTenantID(2)})
+	protectTenants(state, hlc.Timestamp{WallTime: 6}, []roachpb.TenantID{roachpb.MakeTenantID(2)})
+
+	ptsStateReader, err := spanconfigsqltranslator.TestingNewProtectedTimestampStateReader(context.Background(), *state)
+	require.NoError(t, err)
+	require.Equal(t, `[ts=3]`, outputTimestamps(ptsStateReader.GetProtectedTimestampsForCluster()))
```

Comment:
This is fine, but I think it's also fine just avoiding all this string formatting and asserting on the actual values directly.
```
cpts := ptsStateReader.GetProtectedTimestampsForCluster()
require.Len(t, 1, cpts)
require.Equal(t, hlc.Timestamp{WallTime: 3}, cpts)
```

Maybe a helper to define hlc.Timestamp like we do in several places: https://github.com/cockroachdb/cockroach/blob/a8fc111d61d409d1f81608aa60f9e755f0ce8aa5/pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go#L43-L45

Reply:
Ditto for the assertions below. Would we have added these string formatters for test had we not written the datadriven thing the first pass?